In [2]:
import pandas as pd

With `groupby()` we can reshape our dataset

In [3]:
pokemons = pd.read_csv('/home/ashporwal/hdd/lab/programmes/Python/Pandas/4_dataframe_part3/Pokemon.csv', index_col='#')
pokemons

Name   Type 1  Type 2  Total  HP  Attack  Defense  \
#                                                                         
1                Bulbasaur    Grass  Poison    318  45      49       49   
2                  Ivysaur    Grass  Poison    405  60      62       63   
3                 Venusaur    Grass  Poison    525  80      82       83   
3    VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4               Charmander     Fire     NaN    309  39      52       43   
..                     ...      ...     ...    ...  ..     ...      ...   
719                Diancie     Rock   Fairy    600  50     100      150   
719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
#                                                    
1         65       65     45           1      False  
2         80       80     60           1      False  
3        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
719      100      150     50           6       True  
719      160      110    110           6       True  
720      150      130     70           6       True  
720      170      130     80           6       True  
721      130       90     70           6       True  

[800 rows x 12 columns]

Lets take a look at the values with have in `Type 1`

In [4]:
pokemons['Type 1'].value_counts() # These are the values we want to group by

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: count, dtype: int64

in above output we can see we have 112 Water type pokemon, 98 Normal and so on ...

So, Now I will try to group these values with `groupby()`

In [5]:
poke_groups = pokemons.groupby(by='Type 1')

poke_groups # doesn't returns anything just an object

In [6]:
# we can check all the values of above poke_groups variable
poke_groups.groups # This will return group of each Type with the indices as value

{'Bug': [10, 11, 12, 13, 14, 15, 15, 46, 47, 48, 49, 123, 127, 127, 165, 166, 167, 168, 193, 204, 205, 212, 212, 213, 214, 214, 265, 266, 267, 268, 269, 283, 284, 290, 291, 292, 313, 314, 401, 402, 412, 413, 413, 413, 414, 415, 416, 469, 540, 541, 542, 543, 544, 545, 557, 558, 588, 589, 595, 596, 616, 617, 632, 636, 637, 649, 664, 665, 666], 'Dark': [197, 198, 215, 228, 229, 229, 261, 262, 302, 302, 359, 359, 430, 461, 491, 509, 510, 559, 560, 570, 571, 624, 625, 629, 630, 633, 634, 635, 686, 687, 717], 'Dragon': [147, 148, 149, 334, 334, 371, 372, 373, 373, 380, 380, 381, 381, 384, 384, 443, 444, 445, 445, 610, 611, 612, 621, 643, 644, 646, 646, 646, 704, 705, 706, 718], 'Electric': [25, 26, 81, 82, 100, 101, 125, 135, 145, 172, 179, 180, 181, 181, 239, 243, 309, 310, 310, 311, 312, 403, 404, 405, 417, 462, 466, 479, 479, 479, 479, 479, 479, 522, 523, 587, 602, 603, 604, 642, 642, 694, 695, 702], 'Fairy': [35, 36, 173, 175, 176, 209, 210, 468, 669, 670, 671, 682, 683, 684, 685, 700, 7

Bug is the name of the pokemon, but this is not correct way to read the dataframe

In [7]:
poke_groups.get_group('Grass').head() # so, we can get the group by referring the type

Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
#                                                                               
1               Bulbasaur  Grass  Poison    318  45      49       49       65   
2                 Ivysaur  Grass  Poison    405  60      62       63       80   
3                Venusaur  Grass  Poison    525  80      82       83      100   
3   VenusaurMega Venusaur  Grass  Poison    625  80     100      123      122   
43                 Oddish  Grass  Poison    320  45      50       55       75   

    Sp. Def  Speed  Generation  Legendary  
#                                          
1        65     45           1      False  
2        80     60           1      False  
3       100     80           1      False  
3       120     80           1      False  
43       65     30           1      False

groupby with more than 1 column

In [8]:
poke_groups = pokemons.groupby(by=['Type 1', 'Type 2'], dropna=False)
poke_groups.groups

{('Bug', 'Electric'): [595, 596], ('Bug', 'Fighting'): [214, 214], ('Bug', 'Fire'): [636, 637], ('Bug', 'Flying'): [12, 123, 127, 165, 166, 193, 267, 284, 291, 414, 415, 416, 469, 666], ('Bug', 'Ghost'): [292], ('Bug', 'Grass'): [46, 47, 413, 540, 541, 542], ('Bug', 'Ground'): [290, 413], ('Bug', 'Poison'): [13, 14, 15, 15, 48, 49, 167, 168, 269, 543, 544, 545], ('Bug', 'Rock'): [213, 557, 558], ('Bug', 'Steel'): [205, 212, 212, 413, 589, 632, 649], ('Bug', 'Water'): [283], ('Bug', nan): [10, 11, 127, 204, 265, 266, 268, 313, 314, 401, 402, 412, 588, 616, 617, 664, 665], ('Dark', 'Dragon'): [633, 634, 635], ('Dark', 'Fighting'): [559, 560], ('Dark', 'Fire'): [228, 229, 229], ('Dark', 'Flying'): [198, 430, 629, 630, 717], ('Dark', 'Ghost'): [302, 302], ('Dark', 'Ice'): [215, 461], ('Dark', 'Psychic'): [686, 687], ('Dark', 'Steel'): [624, 625], ('Dark', nan): [197, 261, 262, 359, 359, 491, 509, 510, 570, 571], ('Dragon', 'Electric'): [644], ('Dragon', 'Fairy'): [334], ('Dragon', 'Fire'):

Once we have grouped any column, then it is pretty easy to perform operations on that dataframe.

In [9]:
max_poke = poke_groups.max().round(2)
max_poke

Name  Total   HP  Attack  Defense  \
Type 1 Type 2                                                           
Bug    Electric                   Joltik    472   70      77       60   
       Fighting  HeracrossMega Heracross    600   80     185      115   
       Fire                    Volcarona    550   85      85       65   
       Flying                    Yanmega    600   86     155      120   
       Ghost                    Shedinja    236    1      90       45   
...                                  ...    ...  ...     ...      ...   
Water  Poison                 Tentacruel    515   80      95       75   
       Psychic                   Starmie    590   95      75      180   
       Rock                     Tirtouga    495  100     108      133   
       Steel                    Empoleon    530   84      86       88   
       NaN                     Wartortle    770  170     150      120   

                 Sp. Atk  Sp. Def  Speed  Generation  Legendary  
Type 1 Type 2                                                    
Bug    Electric       97       60    108           5      False  
       Fighting       40      105     85           2      False  
       Fire          135      105    100           5      False  
       Flying        116      110    160           6      False  
       Ghost          30       30     40           3      False  
...                  ...      ...    ...         ...        ...  
Water  Poison         80      120    100           2      False  
       Psychic       130      110    115           2      False  
       Rock           83       85     55           5      False  
       Steel         111      101     60           4      False  
       NaN           180      160    115           6       True  

[154 rows x 10 columns]

Now, how we can use `agg()` with group

In [10]:
# we are grouping by Type 1 and performing agg, and in agg we can pass the column
# name and the opertaion we want to perform
pokemons.groupby('Type 1').agg(
    {
        'Total': ['max', 'min', 'mean'] # name and the opertaion we want to perform

    }
).round(2)

Total             
           max  min    mean
Type 1                     
Bug        600  194  378.93
Dark       680  220  445.74
Dragon     780  300  550.53
Electric   610  205  443.41
Fairy      680  218  413.18
Fighting   625  210  416.44
Fire       680  250  458.08
Flying     580  245  485.00
Ghost      680  275  439.56
Grass      630  180  421.14
Ground     770  265  437.50
Ice        580  250  433.46
Normal     720  190  401.68
Poison     535  245  399.14
Psychic    780  198  475.95
Rock       700  280  453.75
Steel      700  300  487.70
Water      770  200  430.46